convert guard_size_oblivious to runtime check in infer_size_impl

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #153432
* #153198
* __->__ #148872

its ok to check the requirement  numel == newsize at runtime in case of unbacked instead of at compile time and assume that its true.

laithsakka committed:
convert guard_size_oblivious to runtime check in infer_size_impl

[ghstack-poisoned]

laithsakka committed:
Update on " convert guard_size_oblivious to runtime check in infer_size_impl"




[ghstack-poisoned]

laithsakka committed:
Update on " [DRAFT] convert guard_size_oblivious to runtime check in infer_size_impl"




[ghstack-poisoned]

laithsakka committed:
Update on " [DRAFT] convert guard_size_oblivious to runtime check in infer_size_impl"




[ghstack-poisoned]

laithsakka committed:
Update on " [DRAFT] convert guard_size_oblivious to runtime check in infer_size_impl"




[ghstack-poisoned]

laithsakka committed:
Update on "convert guard_size_oblivious to runtime check in infer_size_impl"




[ghstack-poisoned]

aorenste commented on aten/src/ATen/InferSize.h:
```diff
@@ -37,31 +39,39 @@ inline void infer_size_impl(
+      numel);
+  if (infer_dim) {
+    set_infer_dim();
+  }
```
The old code ALWAYS threw an exception here but the new code might not... intended?

laithsakka replied:
it did not always throw before there was a return at line 59 in the old code.

laithsakka replied:
the conditions are maintained if you try to decompose and flatten the checks

laithsakka commented:
need to look at test_unflatten

laithsakka committed:
Update on "convert guard_size_oblivious to runtime check in infer_size_impl"


its ok to check the requirement  numel == newsize at runtime in case of unbacked instead of at compile time and assume that its true.

[ghstack-poisoned]

bobrenjc93 commented on aten/src/ATen/InferSize.h:
```diff
@@ -25,6 +25,8 @@ inline void infer_size_impl(
   std::optional<int64_t> infer_dim;
   for (int64_t dim = 0, ndim = shape.size(); dim != ndim; dim++) {
+    // We can avoid failing on unbacked shape[dim] and assert that it is >=0
+    // following python behaviour.
```
I don't see how this comment relates to the code underneath

laithsakka replied:
when shape[dim] is unbacked we will fail when we do the check a later PR 
address this by adding Guard_or_false. 
how do you suggest that i rephrase the comment. ?

laithsakka replied:
I dont there is need for todo, and issue number since next PR is already up and have the change I can remove the comment completly

bobrenjc93 commented on aten/src/ATen/InferSize.h:
```diff
@@ -37,31 +39,39 @@ inline void infer_size_impl(
-    }
+  auto set_infer_dim = [&]() {
+    // We have a degree of freedom here to select the dimension size; follow
+    // NumPy semantics and just bail.  However, a nice error message is needed
```
I found this a bit hard to read. How about

```
    // In the face of dimension size ambiguity, we opt to follow numpy semantics and insert a runtime assert that errors at runtime if we see a tensor with no elements.
```

The whole nice error message part seems self evident?

laithsakka replied:
this is exisitng comment i did not add it. its moved as is

bobrenjc93 replied:
I see, it's fine to land as is I guess. Bonus points if you clean up the comment to be more understandable.

laithsakka replied:
I can do that.

bobrenjc93 reviewed (approved):
please make comments more understandable before landing

laithsakka committed:
Update on "convert guard_size_oblivious to runtime check in infer_size_impl"


its ok to check the requirement  numel == newsize at runtime in case of unbacked instead of at compile time and assume that its true.

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on "convert guard_size_oblivious to runtime check in infer_size_impl"


its ok to check the requirement  numel == newsize at runtime in case of unbacked instead of at compile time and assume that its true.

[ghstack-poisoned]

laithsakka commented:
replaced by https://github.com/pytorch/pytorch/pull/153198

laithsakka commented:
@pytorchmergebot merge

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot merge -i
